### ISSUE TITLE
UBL Menu Sytem

### ISSUE THREAD

**Issue by Tannoo** [Tannoo/Marlin#1]
@Roxy-3D, could you look this over and tell me what you think?

https://github.com/Tannoo/Marlin/tree/UBL_Menu

It is current with RCBugFix as of 4/20/17.

**Comment by Tannoo** [comment]
I would like to make `GRID_MAX_POINTS_X` adjustable. 
Or, at least, UBL grid points adjustable.

**Comment by Roxy-3D** [comment]
I don't understand the question...    The GRID_MAX_POINTS_X & Y need to be known at compile time.  Do you want to make those changeable at print time?   That may not make sense because a lot of storage is allocated based on those numbers and that storage would still have to be allocated for the larger, worst case.

The LCD Menu changes may be very helpful.   That would allow somebody to setup a mesh without a host computer!

**Comment by Tannoo** [comment]
Ok. Don't worry about it. Not a big deal.

I am also trying to figure out how to get the O1 output on the LCD.

Any help there?

**Comment by Roxy-3D** [comment]
I'm not sure what you are thinking...  If you want to have the LCD Panel dump the mesh in a form it can be grabbed by the host computer, it might be reasonable to add something to ultraLCD.cpp similar to how home offsets does its work:
```cpp
    void lcd_set_home_offsets() {
      // M428 Command
      enqueue_and_echo_commands_P(PSTR("M428"));
      lcd_return_to_status();
    }
```
Except you would be doing a PSTR("G29 O1");

Is that what you are asking???

**Comment by Tannoo** [comment]
No, I got that part.

I would like to have the option to view the mesh on the LCD. At least the numerical matrix.
Dump to the LCD, not the serial.

I'm not sure that there is a small enough font to display a 15x15 matrix.
So, it may not be possible, it's just a thought.

**Comment by Tannoo** [comment]
I would like to setup another map_type to output to the LCD an not the serial.

I have been fighting with all kinds of stuff to get it to work.

Maybe you can help me out with the flow of information:
It looks like this to me:

`ubl_g29.cpp` calls `ubl.display_map` when 'O' is seen.
That calls this: `void unified_bed_leveling::display_map` in `ubl.cpp`.

It there, in that function, that things get a little fuzzy to me.

First step I've done is to change the map_type variable to int instead of bool.

My real confusion is this:

Now, trying to figure out how to make the serial echoes into STATIC_ITEMs for the LCD without all the conflicts and non-definitions.

Do I need to make a `START_SCREEN()` in the `ubl.cpp` or somehow get the x,y numbers called from an `ultralcd.cpp` function?

**Comment by Roxy-3D** [comment]
> It there, in that function, that things get a little fuzzy to me.
> First step I've done is to change the map_type variable to int instead of bool.

Yeah...  Thinky did that.  But I don't really agree a Bool is better than an integer map type.  And a Bool limits you to two map types.  So...  I just left it for now.  But for a 3rd map type, you want to change that.

The map function isn't that complicated if you realize, the bulk of it is to print the headers (and trailers) above (and below) the map.  And then the other complexity is it is trying to figure out which format to print in.  Really...  without all that extra complexity it turns into:
```cpp
   for (int8_t j = GRID_MAX_POINTS_Y - 1; j >= 0; j--) {
      for (uint8_t i = 0; i < GRID_MAX_POINTS_X; i++) {
        float f;
	f = z_values[i][j];
        if (isnan(f)) 
          serialprintPGM(PSTR("    .    "));
        else 
          SERIAL_PROTOCOL_F(f, 3);
     }
      MYSERIAL.write((char)'\n');
    }
```
> Now, trying to figure out how to make the serial echoes into STATIC_ITEMs for the LCD without all the conflicts and non-definitions.
> Do I need to make a START_SCREEN() in the ubl.cpp or somehow get the x,y numbers called from an ultralcd.cpp function?

I can't help you much with the LCD code.  That stuff is cryptic.  When I have to do something in that file, I find a function or feature that behaves similarly to what I want to do and copy it.  And then start making small changes to it.    But one way you can see the Mesh values on the LCD is to do a G29 P4 R and it will move the nozzle to each location and display the current mesh point value.  If you click the encoder wheel without turning it...  You can walk through all the numbers in a given area of the mesh.

I know that isn't what you want but at least it does get you the information displayed on the LCD Panel.

**Comment by Tannoo** [comment]
I have a screen setup to display some info... now I just need to find and assign the info to display.

**Comment by Tannoo** [comment]
I believe the magic happens here:
```
    const float current_xi = ubl.get_cell_index_x(current_position[X_AXIS] + (MESH_X_DIST) / 2.0),
                current_yi = ubl.get_cell_index_y(current_position[Y_AXIS] + (MESH_Y_DIST) / 2.0);

    for (int8_t j = GRID_MAX_POINTS_Y - 1; j >= 0; j--) {
      for (uint8_t i = 0; i < GRID_MAX_POINTS_X; i++) {
        const bool is_current = i == current_xi && j == current_yi;

        // is the nozzle here? then mark the number
        if (map_type == 0) SERIAL_CHAR(is_current ? '[' : ' ');

        const float f = z_values[i][j];
        if (isnan(f)) {
          serialprintPGM((map_type == 0) ? PSTR("   .  ") : PSTR("NAN"));
        }
        else {
          // if we don't do this, the columns won't line up nicely
          if ((map_type == 0) && f >= 0.0) SERIAL_CHAR(' ');
          SERIAL_PROTOCOL_F(f, 3);
          idle();
        }
        if ((map_type >= 1) && i < GRID_MAX_POINTS_X - 1) SERIAL_CHAR(',');

        #if TX_BUFFER_SIZE > 0
          MYSERIAL.flushTX();
        #endif
        if (map_type == 0) {
          SERIAL_CHAR(is_current ? ']' : ' ');
          SERIAL_CHAR(' ');
        }
      }
      SERIAL_EOL;
      if (j && (map_type == 0)) { // we want the (0,0) up tight against the block of numbers
        SERIAL_CHAR(' ');
        SERIAL_EOL;
      }
    }
```

**Comment by Tannoo** [comment]
I didn't see your reply when I posted the last two...

Thanks for your help. I will hammer my head on the keyboard until something works. lol

**Comment by Roxy-3D** [comment]
Yes...  That block of code I Cut & Pasted into the above text is the 'magic' to send the Mesh Map to the host device.   Instead of using serial and MYSERIAL routines, you would be calling LCD functions to get things displayed.
```cpp   
for (int8_t j = GRID_MAX_POINTS_Y - 1; j >= 0; j--) {
      for (uint8_t i = 0; i < GRID_MAX_POINTS_X; i++) {
        float f;
	f = z_values[i][j];
        if (isnan(f)) 
          serialprintPGM(PSTR("    .    "));
        else 
          SERIAL_PROTOCOL_F(f, 3);
     }
      MYSERIAL.write((char)'\n');
    }
```

**Comment by Tannoo** [comment]
Yeah.. I'm working on that.

PITA, it is.

**Comment by Tannoo** [comment]
I give up, my head is bleeding.

**Comment by Tannoo** [comment]
I just can't get declarations to jive.

I tried to make it work from `ultralcd.cpp`.... nope... can't get the `f` to carry over. Something about cannot float an int. lol

I cannot get it to work in `ubl.cpp`... nope. That's not happening either. Can't seem to draw a screen from there.

**Comment by Roxy-3D** [comment]
To convert an integer to float...     You just cast it with a    ((float) i)     assuming i is an int.

Somebody else over at the RepRap forum wanted to make changes to the LCD Panel menu code.   I told them this is a very difficult piece of code to be trying to modify if you are not fluent in C.

http://forums.reprap.org/read.php?415,763824

Actually....  I could have said that better.   The LCD Panel code is very difficult to modify EVEN IF you are fluent in C.   It is a very cryptic piece of code.

**Comment by Tannoo** [comment]
Yeah. I don't want to convert anything as that would mean losing the resolution of the z corrections of the matrix displayed on the LCD.

Question: What is the `\` used for in C/C++/Arduino?
```
  #define START_SCREEN() \
    START_SCREEN_OR_MENU(LCD_HEIGHT - (TALL_FONT_CORRECTION)); \
    encoderTopLine = encoderLine; \
    bool _skipStatic = false; \
    SCREEN_OR_MENU_LOOP()
```

**Comment by Roxy-3D** [comment]
It is a pre-processor directive.  It says that the stuff on the next line is still part of the current macro definition.   The START_SCREEN() macro includes everything on the next 4 lines.

**Comment by Tannoo** [comment]
Okay, this compiles... but, doesn't do what's expected.
My guess is memory corruption. Wrong locations somewhere.

The screen doesn't show, but servos are wigging out and after a few seconds, Marlin resets.

This is executed from `ultralcd.cpp`:
```
    void _lcd_ubl_output_lcd_map() {
      START_SCREEN();
        STATIC_ITEM(MSG_UBL_LEVEL_BED, false, false);

        float z_values[GRID_MAX_POINTS_X][GRID_MAX_POINTS_Y];
      
        const float current_xi = ubl.get_cell_index_x(current_position[X_AXIS] + (MESH_X_DIST) / 2.0),
                    current_yi = ubl.get_cell_index_y(current_position[Y_AXIS] + (MESH_Y_DIST) / 2.0);
  
        for (int8_t j = GRID_MAX_POINTS_Y - 1; j >= 0; j--) {
          for (uint8_t i = 0; i < GRID_MAX_POINTS_X; i++) {
            const bool is_current = i == current_xi && j == current_yi;
  
            const float f = z_values[i][j];
            if (isnan(f)) {
              STATIC_ITEM("NAN", false, false);
            }
            else {
              // if we don't do this, the columns won't line up nicely
              STATIC_ITEM("", f, false, false);
              idle();
            }
            if (i < GRID_MAX_POINTS_X - 1) STATIC_ITEM(",", false, false);
  
            safe_delay(15);
          }
          SERIAL_EOL;
        }
      END_SCREEN();
    }
```
The serial output shows something that I didn't expect:
```
save_ubl_active_state_and_disabled() called multiple times in a row.
restore_ubl_active_state_and_leave() called too many times.
echo:Bed Leveling Off
ok
```

**Comment by Tannoo** [comment]
I expected the z_value to be displayed on one line, and each one to display in the same spot for now.

I am trying to get something to start working...after that... formatting.

If I have to ditch displaying the map on the LCD, I can. I just think that it's doable.

**Comment by Roxy-3D** [comment]
The 
```cpp
save_ubl_active_state_and_disabled() called multiple times in a row.
restore_ubl_active_state_and_leave() called too many times.
echo:Bed Leveling Off
ok
```
message is strange.   That isn't even in that code.  That is some extra code to make sure recursion isn't happening and some how that code is tripping it.

Let me think about this and see if something suddenly makes sense to me.

**Comment by Tannoo** [comment]
I do believe the `START_MENU` / `END_MENU` is called several times before it stops. IDK why, but that's what I've seen.
I will put a delay in there to see if that helps.

**Comment by Tannoo** [comment]
To no repeat a function, I found that I can call that one:
```
    void _lcd_ubl_output_map_numbers() {
      ubl.display_map(map_type);
    }
```
I am not trying to get numbers from it. So far, I get no numbers... just random stepper moves. lol .. Not exactly what I'm looking for.

**Comment by Tannoo** [comment]
Kill screen shows in less than 1 sec. Here's the serial output:
```
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.1520.256Error:KILL caused by too much inactive time - current command: 
Error:Printer halted. kill() called!
```

**Comment by Tannoo** [comment]
Here's the relevant functions in `ultralcd.cpp`:
```
    void _lcd_ubl_output_map_numbers() {
      ubl.display_map(map_type);
    }

    /**
     * UBL Output LCD map
     */
    void _lcd_ubl_output_lcd_map() {
      START_SCREEN();
      STATIC_ITEM(MSG_UBL_LEVEL_BED, false, false);
      _lcd_ubl_output_map_numbers();
      STATIC_ITEM("", lcd_output_f, false, false);
      END_SCREEN();
    }
```

**Comment by Tannoo** [comment]
If I add a delay inside the SCREEN calls, I get X, Y, AND Z movements and then Marlin resets.
```
    void _lcd_ubl_output_map_numbers() {
      ubl.display_map(map_type);
    }

    /**
     * UBL Output LCD map
     */
    void _lcd_ubl_output_lcd_map() {
      START_SCREEN();
      STATIC_ITEM(MSG_UBL_LEVEL_BED, false, false);
      _lcd_ubl_output_map_numbers();
      STATIC_ITEM("", lcd_output_f, false, false);
      safe_delay(200);
      END_SCREEN();
    }
```
Here's the serial up to the reset:
```
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152start
echo:Marlin 1.1.0-RCBugFix
```
So, this time, I don't know what reset it or why.

**Comment by Tannoo** [comment]
Is there a way to call a function only once and null after that?

**Comment by Tannoo** [comment]
I'm sorry, I am thinking out loud. I created a `UBL_MAP_CALL_ONCE` flag to hopefully do just that.

**Comment by Tannoo** [comment]
lol.... I changed the code to this:
```
    void _lcd_ubl_output_map_numbers() {
      ubl.display_map(map_type);
      UBL_MAP_CALL_ONCE = false;
    }

    /**
     * UBL Output LCD map
     */
    void _lcd_ubl_output_lcd_map() {
      START_SCREEN();
      STATIC_ITEM(MSG_UBL_LEVEL_BED, false, false);
      if (UBL_MAP_CALL_ONCE == true) _lcd_ubl_output_map_numbers();
      STATIC_ITEM("", lcd_output_f, false, false);
      END_SCREEN();
    }
```
No Screen change, but I get this on the monitor:
```
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.152
Bed Topography Report for LCD:

0.1520.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
0.2560.3120.3020.2620.1550.008-0.2320.2000.200
0.1430.2320.2900.2760.2240.1610.027-0.1990.2000.200
0.0860.2350.2480.2470.2140.1730.047-0.1550.2000.200
0.0590.1710.2100.2090.1890.1560.080-0.1040.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.2000.2000.2000.2000.2000.2000.2000.2000.200
0.2000.1320.1290.1170.1030.0610.019-0.079-0.3030.200
0.2000.2050.2530.2180.2350.1820.135-0.029-0.2620.200
0.2000.1900.2420.2890.2650.2240.136-0.018-0.2680.200
0.2000.1720.2610.3030.3000.2410.1620.005-0.2550.200
Error:MAXTEMP triggered, system stopped! Heater_ID: 0
Error:Printer halted. kill() called!
```
Very interesting. The LCD actually timed out and was at the info screen before it killed it. I'm just not sure why the heater is involved yet. OH! the last time must have set the temps somehow, because everything is  very hot!

**Comment by Tannoo** [comment]
Yes, it's turning on all the analog outputs. The hotend, the bed, and the fan. But with arbitrary values.
Something is getting stepped on.

**Comment by Tannoo** [comment]
Hotend was set to 744, the bed to 1260, and the fan to 12%. Very odd.

**Comment by Roxy-3D** [comment]
Two comments...   You have to call idle() very often or the Kill message
will happen.  That is to protect the printer in case the firmware goes
crazy and doesn't keep  managing the heaters.

The hotend going 744 degrees sounds like you have some data corruption
happening.   A bad pointer or something is stepping on things.   It might
be best to find the data corruption problem first, because it maybe the
rest of your code is working right but you can't tell that with the
corruption happening.

**Comment by Tannoo** [comment]
Yes, I believe that. I'm trying to find out where.
`ubl.display_map(map_type);` is getting called too many times in a row.
The `idle()` call is in the `ubl.display_map` function.

I'm not sure where the data corruption is. I ran into that once with `char XXX [15]` declarations and got that sorted.
But, this time, there is a float that's new.
In `ubl.cpp`: 
`float lcd_output_f;`
 and 
`if (map_type == 2) lcd_output_f = f;`

In `ultralcd.cpp`: 
`extern float lcd_output_f;` 
and 
`STATIC_ITEM("", lcd_output_f, false, false);`

**Comment by Tannoo** [comment]
I will remove that float and see if the corruption and stuff stops.

**Comment by Tannoo** [comment]
That made no difference... hmm. Well, there was a time that it didn't, now it does.

**Comment by Tannoo** [comment]
I've tried so many different ways to get the `f` data to `ultralcd.cpp` and nothing gives me a number.
Latest iteration crashes with Marlin resetting or kill called. It's random.

So, I've given up for now.

**Comment by Tannoo** [comment]
Well, I tried another approach before I giving up.

...and it's back to weird corruptions again.

I am declaring in `ultralcd.cpp`, `extern const float f` because of `const float f` in `ubl.cpp`.

Then, currently trying to display that as `STRINGIFY(f)`. 

I am lost as to why corruption is happening. But, I'll put it to rest for another while.

**Comment by Roxy-3D** [comment]
> I am lost as to why corruption is happening. But, I'll put it to rest for another while.

I know why it is happening...   It took 4 of us and a hardware debugger to figure it out.  This thread goes into all the gory details.  https://github.com/MarlinFirmware/Marlin/issues/6405#issuecomment-297429048

 I'm sorry.  I thought you were working off of the most current code base.   If I had realized that wasn't the case, I would spoke up much earlier.

Any way....   If you move your changes to the current code base, the data corruption should be gone.

**Comment by Roxy-3D** [comment]
I was just thinking...   Would it make sense to leave the map display alone for now?   And instead add a UBL menu to the LCD Panel code.    Maybe an Activate, Deactivate, P0, P1 P3, P4, Save and Load option?   Then people would not need to have a host computer plugged into the printer.  Somebody was just asking for that.

First, those would be a lot simpler and more straight forward to add to the LCD code.  That would help you do the more difficult Map Display.   But the other thing is I finally have some time to think about things.   There is probably a good way to use the limited display space to show the map.   Maybe we have a little postage stamp size icon with a dot on it representing which number (or group of numbers) is being displayed.   And the dot moves around based on how the encoder wheel is turned?

**Comment by Tannoo** [comment]
>This thread goes into all the gory details. MarlinFirmware#6405 (comment)

Yes, I've seen that.

> I thought you were working off of the most current code base. 

I am.

>If you move your changes to the current code base, the data corruption should be gone.

Actually, I stopped trying to display data from `ubl.cpp`. I copied the relevant stuff over to `ultralcd.cpp` in the function to display the map. I stopped having corruption issues, but it's not displaying any numbers.

>Would it make sense to leave the map display alone for now? And instead add a UBL menu to the LCD Panel code. Maybe an Activate, Deactivate, P0, P1 P3, P4, Save and Load option? Then people would not need to have a host computer plugged into the printer. Somebody was just asking for that.

I was thinking that also. I have everything else working... maybe more than you are asking for. lol

Here's the breakdown of the UBL menu I have now:
```
    /**
     * UBL submenu
     * 
     * Prepare
     *  Unified Bed Leveling
     *    Activate UBL
     *    Deativate UBL
     *    1 - Build Mesh
     *      Build PLA Mesh
     *      Build ABS Mesh
     *      Build Custom Mesh
     *        Hotend Temp:
     *        Bed Temp:
     *        Build Custom Mesh
     *      Build Cold Mesh
     *      Continue Bed Mesh
     *    2 - Fill-in NANs
     *      Smart Fill-in
     *      Fill-in Mesh
     *    3 - Validate Mesh
     *      PLA Mesh Validation
     *      ABS Mesh Validation
     *      Custom Mesh Validation
     *        Hotend Temp:
     *        Bed Temp:
     *        Validate Mesh
     *    4 - Edit Mesh
     *      Edit PLA Mesh
     *      Edit ABS Mesh
     *      Edit Custom Mesh
     *        Hotend Temp:
     *        Bed Temp:
     *        Edit Mesh
     *    Mesh Leveling
     *      Mesh Leveling
     *      Grid Mesh Leveling
     *        Side points:
     *        Level Mesh
     *    Mesh Storage
     *      Memory Slot:
     *      Load Bed Mesh
     *      Save Bed Mesh
     *    Output Map
     *      Map Type:
     *      Output Bed Mesh Host / Output Bed Mesh Excel
     *    Output UBL Info
     */
```
That's how I'm keeping mine for now.

Do you want only what you stated?

**Comment by Roxy-3D** [comment]
Your full menu list is great!   I only gave the smaller list because even that is a fair amount of work.  But that would be the minimal amount needed to have a 'Full System'.     Why would a user want to have a PLA mesh and an ABL mesh?   Any mesh should be able to allow any type of filament to be used with it.

>> I thought you were working off of the most current code base.

> I am.

OK.   But when I go to the code in this branch and look at gcode_G29() it doesn't have the __attribute__(optimization(O0))   or what ever the correct syntax is.   That is why I thought you did not have the most current.     No reason to fight problems that have already been found!

**Comment by Tannoo** [comment]
Well, I haven't updated this branch in a week or so. But, my local working copy on my printer is the latest as of yesterday.

**Comment by Tannoo** [comment]
I can update it.

**Comment by Tannoo** [comment]
It is now updated with RCBugFix as of now.

It will take me a few more mins to get the latest menu code I have on to it.

**Comment by Tannoo** [comment]
Ok. It's updated now to what I have on my printer.

FYI: There are other things I have like:
DHT sensor addition
RGB lighting control menu.
Leveling system in use listed on the Printer Info screen.
Initialize Memory (M502 > M500 > M501") on the Control menu.

**Comment by Tannoo** [comment]
They are all on my `current_running` branch.

**Comment by Tannoo** [comment]
Tell me what you want changed, I will.

**Comment by Tannoo** [comment]
Or...you should be able to submit PRs here.

**Comment by Roxy-3D** [comment]
What is your thinking on why there should be a PLA Mesh and an ABS Mesh?   I just dial my mesh in very accurately, and I use it for either type of material.    

Also...  Another suggestion.  A command like G29 J4 would be useful on your menu.  (To tilt the mesh based on the glass positioning.)    But it takes a lot of extra logic to make the LCD Menu allow you to pick the number.   It might be good to have an include file or a set of #define's where the user could pick a number.   You would be able to enque a G29 J command.  But it would be a lot simpler if the user had to pick a number to go with that ahead of time.   Or maybe...  That is exactly what you do.  You allow them to enque a G29 J but the command defaults to the specified G29 J default (which I believe is 3).

**Comment by Tannoo** [comment]
Well, on mine, the bed changes shape when it's heated to 110C vs 50C

**Comment by Tannoo** [comment]
Do you set your mesh up at ABS temps? I can't do that yet due to PLA is all I have atm. So, I have to set it up at PLA temps.

**Comment by Tannoo** [comment]
>A command like G29 J4 would be useful on your menu.

     *    Mesh Leveling
     *      Mesh Leveling <-- 3-Point
     *      Grid Mesh Leveling
     *        Side points:  <-- Custom Points
     *        Level Mesh  <-- Run with Custom Points
 
```
    /**
     * UBL Mesh Leveling submenu
     */
    void _lcd_ubl_mesh_leveling() {
      START_MENU();
      MENU_BACK(MSG_UBL_LEVEL_BED);
      MENU_ITEM(gcode, MSG_UBL_MESH_LEVELING, PSTR("G29 T"));  <-- 3-Point tilting
      STATIC_ITEM(MSG_UBL_GRID_MESH_LEVELING, true);
      MENU_ITEM_EDIT(int3, MSG_UBL_SIDE_POINTS, &SIDE_POINTS, 2, 5);  <-- Set 2 - 5 points
      MENU_ITEM(function, MSG_UBL_MESH_LEVEL, _lcd_ubl_mesh_level_cmd);  <-- Run with Custom points.
      END_MENU();
    }
```
```
    /**
     * UBL Mesh Leveling Command
     */
    void _lcd_ubl_mesh_level_cmd() {
      sprintf_P(SIDE_POINTS_GCODE, PSTR("G29 J%i"), SIDE_POINTS);  <-- Here's your Jx command.
      enqueue_and_echo_command(SIDE_POINTS_GCODE);
    }
```
I'll rename `MSG_UBL_MESH_LEVELING` to `MSG_UBL_3POINT_MESH_LEVELING`.

**Comment by Tannoo** [comment]
>tilt the mesh 

Mesh leveling?

**Comment by Tannoo** [comment]
Or would you like me to rename that to `Mesh Tilting`?

**Comment by Tannoo** [comment]
Do you think that a range of 2 to 5 is sufficient for the mesh _tilting_?

**Comment by Roxy-3D** [comment]
Probably a Range of 3 to 6 is more than sufficient for Mesh Tilting.
Rather than have a PLA Mesh command and an ABS Mesh Command...  It would make sense to allow the user to Specify which mesh slot to save (or load) the mesh.   I use the same mesh for different materials.  But I do have multiple mesh's defined and saved based on which piece of glass I'm using.

**Comment by Tannoo** [comment]
`J #   Grid   *   Perform a Grid Based Leveling of the current Mesh using a grid with n points on a side.`
I got `Mesh Leveling` from that description.

**Comment by Roxy-3D** [comment]
Yes...  This is actually a combination of the Mesh Leveling idea with the planar Grid Leveling.   The mesh might be perfectly tuned for a piece of glass.  But if a small piece of plastic ends up under the glass...  That mesh will be slightly tilted.

**Comment by Tannoo** [comment]
>Rather than have a PLA Mesh command and an ABS Mesh Command... It would just make sense to allow the user to Specify which mesh slot to save (or load) the mesh.

I can remove the PLA and ABS options.
Leave the Custom? to set desired temps? Or remove that also?

Selecting a storage slot and loading or saving to that slot is already there.
Was there something else you had in mind?
```
    /**
     * UBL Mesh Storage submenu
     */
    void _lcd_ubl_storage_mesh() {
      START_MENU();
      MENU_BACK(MSG_UBL_LEVEL_BED);
      MENU_ITEM_EDIT(int3, MSG_UBL_STORAGE_SLOT, &UBL_STORAGE_SLOT, 0, 9);  <--Slot Select
      MENU_ITEM(function, MSG_UBL_LOAD_MESH, _lcd_ubl_load_mesh); <-- go load it.
      MENU_ITEM(function, MSG_UBL_SAVE_MESH, _lcd_ubl_save_mesh);  <-- go save it.
      END_MENU();
    }
```
```
    /**
     * UBL Load Mesh Command
     */
    void _lcd_ubl_load_mesh() {
      sprintf_P(UBL_STORAGE_GCODE, PSTR("G29 N L%i"), UBL_STORAGE_SLOT);  <-- Load selected slot
      enqueue_and_echo_command(UBL_STORAGE_GCODE);
    }

    /**
     * UBL Save Mesh Command
     */
    void _lcd_ubl_save_mesh() {
      sprintf_P(UBL_STORAGE_GCODE, PSTR("G29 N S%i"), UBL_STORAGE_SLOT);  <-- Save selected slot
      enqueue_and_echo_command(UBL_STORAGE_GCODE);
    }
```

**Comment by Tannoo** [comment]
>This is actually a combination of the Mesh Leveling idea with the planar Grid Leveling. The mesh might be perfectly tuned for a piece of glass. But if a small piece of plastic ends up under the glass... That mesh will be slightly tilted.

I understand. Did you want it to stay as `Mesh Leveling` or title it to something else?

**Comment by Roxy-3D** [comment]
> I can remove the PLA and ABS options.
> Leave the Custom? to set desired temps? Or remove that also?

I understand now why you think that is important.   I believe it makes sense to let the user probe the bed (or for that matter do a G26) at different temperatures.   Maybe we could have a common routine that gets the bed and nozzle temperature, and then uses that to enque the G29 P1 or the G26 commands?

But also...  For the bed and nozzle heating...  There are already #defines for PLA and ABS.  So using those would be just fine.  The more I think about it...   Having the same command with PLA or ABS on the title is just fine.   People will instantly know what is what.

> Selecting a storage slot and loading or saving to that slot is already there.
> Was there something else you had in mind?

I think what you are doing is good!   People are going like using your LCD Menu System and they are not going to have trouble getting it to do the right thing for them.

> Did you want it to stay as Mesh Leveling or title it to something else?

For G29 J,   Probably Mesh Tilting is the correct thing to call it.

**Comment by Tannoo** [comment]
I value your insight.
Just trying to understand what you would like to see.

The generate pla mesh is for preheating before probing to pla temps. -- yes, I should get those temps from the system defines instead of the set temps I placed.
Ditto for generate ABS mesh.

If I were to consolidate those to just one generate mesh, what temps to use? PLA or ABS.

I have found that when doing the UBL probing @ ABS temps with PLA in the hotend, it is way to hot for too long due to no extrusion taking place. This could result in clogging.

**Comment by Roxy-3D** [comment]
> If I were to consolidate those to just one generate mesh, what temps to use? PLA or ABS.

The more I think about it...   It is more intuitive for the user to have both specified and available.

> I have found that when doing the UBL probing @ ABS temps with PLA in the hotend, it is way to hot for too long due to no extrusion taking place. This could result in clogging.

Yeah...   It is more accurate to do your probing at full temperature.   But in that case...  Starting with a PLA grid and doing a G26 at ABS temperatures (using ABS) would work fine.   You would then edit the mesh to correct some locations.   And maybe have to do another iteration to get things perfect.   But you don't have to do the probing with the nozzle at full temperature to end up with a perfect mesh.

But there isn't going to be any way to communicate that on the LCD Panel.  Probably this type of 'Helpful Hint' would go into the Quick Start guide that Bob-the-Kuhn is doing.

**Comment by Tannoo** [comment]
Okay.

**Comment by Tannoo** [comment]
I think I have conserved some memory by using the same variable for gcodes. I will be testing this to make sure all is still kosher.

**Comment by Tannoo** [comment]
This doesn't work like it used to. It now runs the first command and stops.
```
      sprintf_P(CUSTOM_MESH_GCODE, PSTR("G28\nM190 S%i\nM109 S%i\nG29 P1\nM104 S0\nM140 S0"), PREHEAT_1_TEMP_BED, PREHEAT_1_TEMP_HOTEND);
      enqueue_and_echo_command(CUSTOM_MESH_GCODE);
```
I have to split it up like this to work:
```
      enqueue_and_echo_command("G28");
      sprintf_P(CUSTOM_MESH_GCODE, PSTR("M190 S%i"), PREHEAT_1_TEMP_BED);
      enqueue_and_echo_command(CUSTOM_MESH_GCODE);
      sprintf_P(CUSTOM_MESH_GCODE, PSTR("M109 S%i"), PREHEAT_1_TEMP_HOTEND);
      enqueue_and_echo_command(CUSTOM_MESH_GCODE);
      enqueue_and_echo_command("G29 P1");
      enqueue_and_echo_command("M104 S0");
      enqueue_and_echo_command("M140 S0");
```
I will try an older copy of Marlin to try and see what changed. (Because, I know a LOT has changed in a few days now.)

**Comment by Roxy-3D** [comment]
Yeah...   It definitely claims that is possible!  Are you calling drain_injected_commands_P() ?   
Probably you should start an issue and point to these comments.
```cpp
/**
 * Inject the next "immediate" command, when possible, onto the front of the queue.
 * Return true if any immediate commands remain to inject.
 */
static bool drain_injected_commands_P() {
  if (injected_commands_P != NULL) {
    size_t i = 0;
    char c, cmd[30];
    strncpy_P(cmd, injected_commands_P, sizeof(cmd) - 1);
    cmd[sizeof(cmd) - 1] = '\0';
    while ((c = cmd[i]) && c != '\n') i++; // find the end of this gcode command
    cmd[i] = '\0';
    if (enqueue_and_echo_command(cmd))     // success?
      injected_commands_P = c ? injected_commands_P + i + 1 : NULL; // next command or done
  }
  return (injected_commands_P != NULL);    // return whether any more remain
}

/**
 * Record one or many commands to run from program memory.
 * Aborts the current queue, if any.
 * Note: drain_injected_commands_P() must be called repeatedly to drain the commands afterwards
 */
void enqueue_and_echo_commands_P(const char* pgcode) {
  injected_commands_P = pgcode;
  drain_injected_commands_P(); // first command executed asap (when possible)
}
```

**Comment by Tannoo** [comment]
The Serial monitor will show:
```
enqeueing "G28
M190 S50
M109 S195
G29 P1
M104 S0
M140 S0"
```
But, it runs G28 and stops.

**Comment by Roxy-3D** [comment]
>> But, it runs G28 and stops.

> Are you calling drain_injected_commands_P() ?

**Comment by Tannoo** [comment]
```
sprintf_P(CUSTOM_MESH_GCODE, PSTR("M190 S%i\nM109 S%i\nG29 P1\nM104 S0\nM140 S0"), PREHEAT_2_TEMP_BED, PREHEAT_2_TEMP_HOTEND);
enqueue_and_echo_command(CUSTOM_MESH_GCODE);
```

**Comment by Roxy-3D** [comment]
Yeah...  But I think it wants you to follow up with something like:
```cpp
while (drain_injected_commands_P())
  idle();
```

**Comment by Tannoo** [comment]
is `sprintf_P` a `drain_injected_commands_P`?

**Comment by Tannoo** [comment]
Or, is that what the `_P` is for?

**Comment by Roxy-3D** [comment]
No.   sprintf_P()   almost certainly is a 'Program Memory' version of sprintf().   Probably (without looking it up), it is taking the format string out of 'Program Memory'.     It has very little to do with things like drain_injected_commands_P();

And I just noticed another issue...     drain_injected_commands_P();  is trying to take the commands it enque()'s out of program memory.  That is what the _P means.  It uses  strncpy_P(cmd, injected_commands_P, sizeof(cmd) - 1);   which goes into program memory to get the string.   But your sprintf_P() put the results of the string formatting operation into RAM memory.   Something needs to change.  Probably there is an enque() command that does not have a _P on the end of it.

**Comment by Tannoo** [comment]
ok.
I have found that this: `MENU_ITEM(gcode, MSG_UBL_BUILD_PLA_MESH, PSTR("G28\nM190 S50\nM109 S195\nG29 P1\nM104 S0\nM140 S0"));` works just fine with the results like:
```
enqueueing "G28"
enqueueing "M190 S50"
enqueueing "M109 S195"
enqueueing "G29 P1"
enqueueing "M104 S0"
enqueueing "M109 S0"
```
I have found that I can't get the `sprint_P` to do that.

**Comment by Roxy-3D** [comment]
The PSTR("---") accesses a character string that is located in Program Memory.   sprintf() and sprintf_P() are going to write their formatted output into a character string located in RAM memory...

**Comment by Tannoo** [comment]
>enqueue_and_echo_command(CUSTOM_MESH_GCODE);

So, is this my problem? Not using the `enqueue_and_echo_commands_P`?

**Comment by Tannoo** [comment]
I'm not thinking so.. because the new results aren't promising:
```
echo:enqueueing " "
echo:Unknown command: ""
```

**Comment by Roxy-3D** [comment]
The root of the problem is the AVR is a "Harvard Architecture".   The program data space overlaps the RAM memory data space.   You have to make sure the pointer and functions you are using are accessing the right data space.

If something has a _P on it...  That means it is going to try to get a character string out of program memory.   And program memory can only be altered by loading Arduino, recompiling the firmware, and uploading it.

I know it is confusing...     And that is what that PSTR() thing is.   That is a macro that forces what ever string you put inside it to go into program memory instead of eating up RAM memory.

**Comment by Tannoo** [comment]
Ok. I am trying to get something work now.
I'll split it up for now.

**Comment by Tannoo** [comment]
Maybe `sprintf_P` just can't handle that much.

**Comment by Roxy-3D** [comment]
> Maybe sprintf_P just can't handle that much.

No...  I think what is happening is sprintf_P() is taking its format string out of program memory.  But the output from that function is ending up in RAM memory.   (You can't change program memory from within the program.  The only place it can put its results is into RAM memory.)

If you are going to enque() something that is in RAM, you can't be using an enque_P() type of command.

**Comment by Tannoo** [comment]
M'kay.

**Comment by Tannoo** [comment]
I'm using this instead and it works like it's supposed to.
```
MENU_ITEM(gcode, MSG_UBL_BUILD_PLA_MESH, PSTR("G28\nM190 S" STRINGIFY(PREHEAT_1_TEMP_BED) 
  "\nM109 S" STRINGIFY(PREHEAT_1_TEMP_HOTEND) "\nG29 P1\nM104 S0\nM140 S0"));
```

**Comment by Roxy-3D** [comment]
Great!   The only problem with that is you can't change it once the firmware is loaded...  The user is stuck with what ever it was when the firmware is built.   But still...   That is going to be just fine!!!   This is going to be super helpful to people!!!

**Comment by Tannoo** [comment]
That was for the PLA mesh build.

This is for custom temps:
```
    void _lcd_ubl_custom_mesh() {
      START_MENU();
      MENU_BACK(MSG_UBL_LEVEL_BED);
      MENU_ITEM_EDIT(int3, MSG_UBL_CUSTOM_HOTEND_TEMP, &CUSTOM_HOTEND_TEMP, EXTRUDE_MINTEMP, HEATER_0_MAXTEMP);
      MENU_ITEM_EDIT(int3, MSG_UBL_CUSTOM_BED_TEMP, &CUSTOM_BED_TEMP, BED_MINTEMP, BED_MAXTEMP);
      MENU_ITEM(gcode, MSG_UBL_BUILD_CUSTOM_MESH, PSTR("G28\nM190 S" STRINGIFY(CUSTOM_BED_TEMP)
        "\nM109 S" STRINGIFY(CUSTOM_HOTEND_TEMP) "\nG29 P1\nM104 S0\nM140 S0"));
      END_MENU();
    }
```

**Comment by Tannoo** [comment]
I have a question though..

Is `HEATER_0_MAXTEMP` suffice for the active extruder? Or is leveling only active for extruder 0?

**Comment by Roxy-3D** [comment]
Right now...  The G26 stuff assumes it will use extruder 0.  Of course..  eventually...  it will support what ever extruder is on the machine.

> Is HEATER_0_MAXTEMP suffice for the active extruder? 

I don't think you can use this...  If you set the temperature to this number...  even a .1 degree over shoot will cause Kill() to be called.   And HEATER_0_MAXTEMP is going to be a bigger number than what ever actually gets used.   It is a hard limit to shut down the printer if it ever happens.

**Comment by Tannoo** [comment]
Ok. What should I use instead?

**Comment by Roxy-3D** [comment]
In Configuration.h there are now values for:
```cpp

// Preheat Constants
#define PREHEAT_1_TEMP_HOTEND 200
#define PREHEAT_1_TEMP_BED     70
#define PREHEAT_1_FAN_SPEED     0 // Value from 0 to 255

#define PREHEAT_2_TEMP_HOTEND 240
#define PREHEAT_2_TEMP_BED    110
#define PREHEAT_2_FAN_SPEED     0 // Value from 0 to 255
```
Probably these are the values that should be in the 'STRINGIFY()' thing....

They are not called out as PLA and ABS.   But I think things are sorting out such that 1 will be PLA and 2 will be ABS.

**Comment by Tannoo** [comment]
Yup. I got those. 

For custom temps, one may want a higher temp than those.

**Comment by Roxy-3D** [comment]
OK...   But here is the thing...  If you are going to let the user dial in a temperature...   You have to enque() the command without an _P functions in play.   I would have to look at the enque() stuff to know for sure.  My guess is there is one that will work with RAM based command strings.   But if there isn't...   It isn't going to be hard to make one that does the same thing except it takes its string from RAM.   After all...
```cpp
static bool drain_injected_commands_P() {
  if (injected_commands_P != NULL) {
    size_t i = 0;
    char c, cmd[30];
    strncpy_P(cmd, injected_commands_P, sizeof(cmd) - 1);
    cmd[sizeof(cmd) - 1] = '\0';
    while ((c = cmd[i]) && c != '\n') i++; // find the end of this gcode command
    cmd[i] = '\0';
    if (enqueue_and_echo_command(cmd))     // success?
      injected_commands_P = c ? injected_commands_P + i + 1 : NULL; // next command or done
  }
  return (injected_commands_P != NULL);    // return whether any more remain
}
```
That is what the  strncpy_P(cmd, ...) is doing.  It is copying the command that it has been told to enque() out of program memory and moving it to RAM.   So you would just skip that step because the stuff you want to do is already in RAM.   Does that make sense?

**Comment by Tannoo** [comment]
Yes.

**Comment by Tannoo** [comment]
Well, like you stated, the custom data is not working.
```
MENU_ITEM(gcode, MSG_UBL_BUILD_CUSTOM_MESH, PSTR("G28\nM190 S" STRINGIFY(CUSTOM_BED_TEMP)
        "\nM109 S" STRINGIFY(CUSTOM_HOTEND_TEMP) "\nG29 P1\nM104 S0\nM140 S0"));
```

**Comment by Tannoo** [comment]
>Maybe we have a little postage stamp size icon with a dot on it representing which number (or group of numbers) is being displayed. And the dot moves around based on how the encoder wheel is turned?

I like that idea! BUT, I am not the dude for that job. :(

**Comment by Roxy-3D** [comment]
>> Maybe we have a little postage stamp size icon with a dot on it representing which number (or group of numbers) is being displayed. And the dot moves around based on how the encoder wheel is turned?

> I like that idea! BUT, I am not the dude for that job. :(

Not to worry...  Once people start using the LCD Panel as the interface...  Somebody will jump in and say "We need the topology map on the LCD too."

**Comment by Tannoo** [comment]
Did you take out this description on purpose? I use it in the menu options because not all options require homing.
```
   *   N    No Home     G29 normally insists that a G28 has been performed. You can over rule this with an
   *                    N option. In general, you should not do this. This can only be done safely with
   *                    commands that do not move the nozzle.
```

**Comment by Tannoo** [comment]
>Once people start using the LCD Panel as the interface...

The UBL menu will evolve. I also believe that as UBL becomes more automated, this menu will also shrink and may also be nothing more than an option to start it as ABL is currently.

**Comment by Roxy-3D** [comment]
Yes.   N is a special letter in GCode to say what line was transmitted.  It causes problems with Repetier Host.   And, pretty much, users shouldn't be using it.   It was only in there for me (and now you).  So...  I took it out of the description just so there would be less controversy.

**Comment by Tannoo** [comment]
Nevermind

**Comment by Tannoo** [comment]
PR has now been submitted. #6566

**Comment by Tannoo** [comment]
Let the critics begin. lol

**Comment by Roxy-3D** [comment]
Yeah...  You know the drill...   No matter how good, how clean, how elegant it is...

Somebody will say "That's nice but what I really need is...."     Or somebody else will say "If you changed this data type to this, it would be much better.   I wouldn't have made that mistake."
